File: java/ankit_kanyal/devnest_problems/NextPermutation.java
// link- https://devsnest.in/algo-challenges/next-permutation

static int[] solve(int n, int[] nums){
//CODE HERE 
int cur=n-2;
if(n==1) return nums;
while(cur>=0 && nums[cur]>=nums[cur+1]){
    cur--;
}

if(cur>=0){
        int j=nums.length-1;
        while(nums[j]<=nums[cur]) j--;
        swap(nums,cur,j);
    }
    reverse(nums,cur+1,nums.length-1);

return nums;
}

static void swap(int []nums,int first,int second){
    int temp=nums[first];
    nums[first]=nums[second];
    nums[second]=temp;
}

static void reverse(int [] nums,int first,int second){
    while(first<second){
        swap(nums,first++,second--);
    }
}

File: python/vedant_bhatnagar/DevsnestProblems_THA/Phase 1/NextPermutation.py
def solve(n, nums):
    # CODE HERE
    curr = n-2

    while(curr>-1):
        if nums[curr] > nums[curr+1]:
            curr = curr-1
        else:
            break
    
    if curr == -1:
        i = 0
        j = n-1
        while(i<=j):
            nums[i], nums[j] = nums[j], nums[i]
            i = i+1
            j = j-1
        return nums

    if curr > -1:
        for i in range(n-1, curr, -1):
            if nums[i] > nums[curr]:
                nums[i], nums[curr] = nums[curr], nums[i]
                break

        i=curr+1
        j=n-1
        while(i<=j):
            nums[i], nums[j] = nums[j], nums[i]
            i=i+1
            j=j-1
        return nums